src/main/java/com/atlassian/bitbucket/jenkins/internal/status/BitbucketBuildStatusFactoryImpl.java

```diff
@@ -0,0 +1,90 @@
+package com.atlassian.bitbucket.jenkins.internal.status;
+
+import com.atlassian.bitbucket.jenkins.internal.model.BitbucketBuildStatus;
+import com.atlassian.bitbucket.jenkins.internal.model.BuildState;
+import com.atlassian.bitbucket.jenkins.internal.model.TestResults;
+import hudson.model.ItemGroup;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.tasks.junit.TestResultAction;
+import jenkins.branch.MultiBranchProject;
+import org.jenkinsci.plugins.displayurlapi.DisplayURLProvider;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collection;
+
+public final class BitbucketBuildStatusFactoryImpl implements BitbucketBuildStatusFactory {
+
+    private static final Collection<Result> successfulResults = Arrays.asList(Result.SUCCESS, Result.UNSTABLE);
+
+    private final DisplayURLProvider displayURLProvider;
+
+    public BitbucketBuildStatusFactoryImpl() {
+        this(DisplayURLProvider.get());
+    }
+
+    BitbucketBuildStatusFactoryImpl(DisplayURLProvider displayURLProvider) {
+        this.displayURLProvider = displayURLProvider;
+    }
+
+    @Override
+    public BitbucketBuildStatus createLegacyBuildStatus(Run<?, ?> build) {
+        return fromBuild(build, false);
+    }
+
+    @Override
+    public BitbucketBuildStatus createRichBuildStatus(Run<?, ?> build) {
+        return fromBuild(build, true);
+    }
+
+    private BitbucketBuildStatus fromBuild(Run<?, ?> build, boolean isRich) {
+        Job<?, ?> job = build.getParent();
+        ItemGroup parent = job.getParent();
+        boolean isMultibranch = parent instanceof MultiBranchProject;
+
+        String name = isMultibranch ? parent.getDisplayName() + " » " + job.getDisplayName() : job.getDisplayName();
+
+        String key = job.getFullName();
+        String url = displayURLProvider.getRunURL(build);
+        BuildState state;
+        if (build.isBuilding()) {
+            state = BuildState.INPROGRESS;
+        } else if (successfulResults.contains(build.getResult())) {
+            state = BuildState.SUCCESSFUL;
+        } else {
+            state = BuildState.FAILED;
+        }
+        BitbucketBuildStatus.Builder bbs = new BitbucketBuildStatus.Builder(key, state, url)
+                .setName(name)
+                .setDescription(state.getDescriptiveText(build.getDisplayName(), build.getDurationString()));
+
+        if (isRich) {
+            BitbucketRevisionAction revisionAction = build.getAction(BitbucketRevisionAction.class);
+
+            bbs.setBuildNumber(build.getId())
+                    .setTestResults(getTestResults(build))
+                    .setParent(isMultibranch ? parent.getFullName() : job.getFullName());
```

Comment:
I'm not totally sure what strings `parent.getFullName()` and `job.getFullName()` will return, and instead of testing it I thought I'd explain what the parent key is and how it will be used. Because it would be easy to understand the parent key to be a project or something like that.

I'm going to use Bamboo as an example here. So the `parent` is intended to be used by something like our required-builds-extended merge check, where you choose a plan and expect a run of that plan to be green before merging. Sounds simple, but it isn't and here is why.

Bamboo when creating plans for branches uses a new plan key; for example if a plan is configured with key `PRBUILD`, plan branches will have key `PRBUILD0`, `PRBUILD1`, and so on. When a user chooses which build they want to ensure is green they will typically choose the main/parent plan; in this case `PRBUILD`. However when Bamboo reports the build result for a branch the key will be something like `PRBUILD0`.

In the required-builds-extended-plugin we assume Bamboo and thus just strip off the trailing numbers and it works. Though this does mean you can't have a plan name with numbers in it. For example POSTGRES96 and POSTGRES97 may be two different plans for versions 9.6 and 9.7 or PostgreSQL, but they would be treated as one plan on stash.a.c currently because we would truncate the key to POSTGRES.

So for Bamboo obviously we want the `parent` value to be `PRBUILD` and `POSTGRES96`, etc. in the above examples.

For a non-multibranch build the above looks like it must be correct, since we set `key` and `parent` to the same thing, and I've seen how that type of build works to know it isn't creating sub-jobs for each branch. For `MultiBranchProject` it looks correct, but I haven't checked to be sure.

Reply:
First, just clarifying the nomenclature here:
- "getName" is the raw name of an individual item, like the build or project name (e.g. "feature/branch", "MutliBranch Job", "Job Folder")
- "getFullName" is the name of the item and all parent items appended with `/` characters
- "getFullDisplayName" is the same as full name but instead appends them with " > "
We use full name to disambiguate jobs that might have the same name but have folders for parents, but for the simple case, Jenkins is the parent so a Freestyle build's key and parent are identical (the job name).
The choice of parent name gives the Multibranch Project name in Jenkins as opposed to the job you'd run- the job name is that of the branch, so you'd need to the parent alongside the key to get a green build out of Jenkins. That suggests this is incorrect, and we should be using the job name for both projects- which would return something like "Folder/Project/feature-branch"

Reply:
I'm happy to spin up Jenkins and check the results if it would help.
Looking at this it makes me think "parent" isn't a great name. Something like "canonical-key" might have been better.

Reply:
Followed this up with @khughes-atlassian and looks like the branch name should not appear in the parent, so if you're okay with that, I will leave this as-is (feel free to test to make sure that's working as intended though- some examples appear in `BitbucketbuildStatusFactoryImplTest`)